android: Implement secondary display support

DavidRGriswold opened the pull request:
Enable support for a second screen on android, either through a physical connection or using screen mirroring through Chromecast or Miracast. 

### What Currently Works

* A "Secondary Screen Layout" option has been added to the Settings menu on Android where user can choose between (1) System Default (mirroring) (2) Top Screen Only (3) Bottom Screen Only (4) Side by Side. 
* A secondary window is always enabled at emulation start - either on an invisible virtual display or a physical display if it exists. If the setting is "System Default (mirroring)" the virtual display is used instead
* It is possible to switch back and forth between the virtual display and physical display using the settings above

### Issues and changes yet to be done

- [x] Test test test


Addresses #351

DavidRGriswold committed:
Enable the SecondScreenPresentation class

DavidRGriswold committed:
Update everything to enable second screen on android under GL and Vulkan. Still some issues!

DavidRGriswold committed:
Some attempts to enable surface changes

DavidRGriswold committed:
OpenGL is working on surface change, vulkan still no

DavidRGriswold committed:
release surfaces (also fixed vulkan?)

DavidRGriswold committed:
added and enabled layout setting

DavidRGriswold committed:
resolve merge conflict

DavidRGriswold committed:
rearrange switch cases to satisfy linux compiler

DavidRGriswold committed:
openGL is working!

DavidRGriswold committed:
several vk changes to try to fix crashes

DavidRGriswold committed:
maybe vulkan is working?

DavidRGriswold committed:
removing unnecessary code attempts

DavidRGriswold committed:
Simplified secondscreen for better performance

OpenSauce04 committed:
vk_platform.cpp: Fixed build failure caused by bad rebase

OpenSauce04 committed:
vk_present_window.h: Removed stray newline

OpenSauce04 committed:
Applied clang-format

DavidRGriswold committed:
bug fix for odin 2

DavidRGriswold committed:
Merge branch 'master' into secondscreen

OpenSauce04 committed:
Merge branch 'master' into secondscreen

OpenSauce04 committed:
Applied clang-format

OpenSauce04 committed:
Updated license headers

OpenSauce04 committed:
Moved SecondScreen class to org.citra.citra_emu.display

OpenSauce04 commented on src/android/app/src/main/jni/emu_window/emu_window_gl.cpp:
```diff
@@ -204,14 +211,14 @@ void EmuWindow_Android_OpenGL::TryPresenting() {
-    }
+    //    if (presenting_state != PresentingState::Running) [[unlikely]] {
+    //        return;
+    //    }
```
Why was this commented out?

DavidGriswoldTeacher replied:
an attempted debug for some early crashing, pretty sure it didn't matter, will put it back in as a sanity check

OpenSauce04 commented on src/android/app/src/main/jni/native.cpp:
```diff
@@ -151,27 +169,39 @@ static Core::System::ResultStatus RunCitra(const std::string& filepath) {
     Core::System& system{Core::System::GetInstance()};
 
     const auto graphics_api = Settings::values.graphics_api.GetValue();
+    EGLContext* c;
```
Please give this variable a more descriptive name

OpenSauce04 commented on src/android/app/src/main/jni/native.cpp:
```diff
@@ -262,6 +293,7 @@ static Core::System::ResultStatus RunCitra(const std::string& filepath) {
             std::unique_lock pause_lock{paused_mutex};
             running_cv.wait(pause_lock, [] { return !pause_emulation || stop_run; });
             window->PollEvents();
+            // if (second_window) second_window->PollEvents();
```
Why was this commented out code added?

OpenSauce04 committed:
Various formatting and readability improvements

OpenSauce04 committed:
Added brackets where previously absent for readability

OpenSauce04 commented on src/android/app/src/main/jni/emu_window/emu_window.cpp:
```diff
@@ -49,16 +49,19 @@ void EmuWindow_Android::OnFramebufferSizeChanged() {
-
+    if (is_secondary) {
+        LOG_DEBUG(Frontend, "Initializing secondary window Android");
+    }
```
This debug message doesn't really make sense. The `EmuWindow_Android` being referred to in the above message is a class of the same name, not a window. It's initializing the values for the `EmuWindow_Android` class.

Mentioning that the secondary window is being initialized in a place where that isn't happening seems incorrect.

OpenSauce04 committed:
Additional readability improvement

OpenSauce04 committed:
RendererVulkan::NotifySurfaceChanged: Simplified condition checking

OpenSauce04 commented on src/video_core/renderer_vulkan/renderer_vulkan.cpp:
```diff
@@ -895,6 +895,19 @@ void RendererVulkan::SwapBuffers() {
+#ifdef ANDROID
+    if (secondary_window) {
+        const auto& secondary_layout = secondary_window->GetFramebufferLayout();
+        if (!second_window) {
```
I know this isn't your fault, but the presence of a variable called both `secondary_window` and `second_window` in the same scope which have different values is really dumb.

We should probably address that here while we're touching this code.

DavidRGriswold replied:
my limited pushback here is that the two equivalent variables for the primary window are "window" and "main_window" which are also pretty confusable! If I go to change one, should I also change the other?

OpenSauce04 replied:
> my limited pushback here is that the two equivalent variables for the primary window are "window" and "main_window" which are also pretty confusable! If I go to change one, should I also change the other?

Yeah it's probably a good idea to change both

OpenSauce04 commented on src/android/app/src/main/jni/native.cpp:
```diff
@@ -304,28 +333,68 @@ extern "C" {
+                                                                     jobject surf) {
+    auto& system = Core::System::GetInstance();
+
+    if (s_secondary_surface) {
+        ANativeWindow_release(s_secondary_surface);
+        s_secondary_surface = nullptr;
+    }
+    s_secondary_surface = ANativeWindow_fromSurface(env, surf);
+    if (!s_secondary_surface) {
+        return;
+    }
```
Is it even possible for this bottom `if` conditional to be true?

DavidRGriswold reviewed (commented):
> Is it even possible for this bottom `if` conditional to be true? 

I believe that `ANativeWindow_fromsurface(env,surf)` can return a nullptr value that would evaluate as false here, but I'm admittedly pretty iffy on some c++ stuff like this

DavidRGriswold reviewed (commented):
> I know this isn't your fault, but the presence of a variable called both `secondary_window` and `second_window` in the same scope which have different values is really dumb.
> 
> We should probably address that here while we're touching this code. 

I am not going to disagree! I would be happy to see those variable names changed to something more reasonable.

DavidRGriswold reviewed (commented):
> This debug message doesn't really make sense. The `EmuWindow_Android` being referred to in the above message is a class of the same name, not a window. It's initializing the values for the `EmuWindow_Android` class.
> 
> Mentioning that the secondary window is being initialized in a place where that isn't happening seems incorrect. 

Fair, should I just remove the debug message at all, or write "Initializing EmuWindow_Android for secondary display"? Definitely a case where I was just trying for parallelism.

DavidRGriswold reviewed (commented):
> Why was this commented out code added? 

an assumption that it was necessary for parallel relationship with the first one, I commented it out at some point in a debugging spree and it seemed to make no difference so I guess I should just remove it.

OpenSauce04 committed:
Merge branch 'master' into secondscreen

DavidRGriswold committed:
change all references to "secondary screen" to "secondary display" to limit confusion with top screen / bottom screen

DavidRGriswold committed:
rename main_window to main_present_window and second_window to secondary_present_window

OpenSauce04 committed:
Reverted accidentally downgraded compatibility list submodule

OpenSauce04 committed:
Removed unnecessary log message

OpenSauce04 committed:
Applied clang-format

OpenSauce04 committed:
Merge branch 'master' into secondscreen

OpenSauce04 committed:
Added a description to the Secondary Display Screen Layout setting

OpenSauce04 committed:
Added `_ptr` suffix to `secondary_present_window`

This distinguishes it as a pointer, as `main_present_window` isn't a pointer, so there could be confusion on whether to use `.` or `->`

OpenSauce04 reviewed (approved):
Looks good to me.
Thanks for implementing this!
